### PR TITLE
Break some more `knative/build` dependencies

### PR DESCRIPTION
**Pull request by vdemeester**
# Changes

Part of #252 but not complete yet — what this does : 

- Remove `GetBuildSpec` from `task_types, and thus remove `knative/build` import from `pkg/apis` package
- Remove using creating a `BuildStatus` from `Pod` to then convert that to a `TaskRunStatus`, let's be more direct.

cc @ImJasonH 

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md#principles) (if functionality changed/added)
- <del>[ ] Includes [docs](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md#principles) (if user facing)</del>
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md)
for more details._

### CODE REVIEW

**Comment by pivotal-nader-ziada** [comment]
Are you planning on making the other changes in the issue in other PRs?

**Comment by vdemeester** [comment]
@pivotal-nader-ziada if this one looks good and is green, I'll open a follow-up PR (as it can be done separately). I just wanted to have a "not-too-big" PR :angel:

**Comment by shashwathi** [comment]
These changes are good to go by itself too so follow up PR for further changes SGTM.

File: pkg/apis/pipeline/v1alpha1/task_validation.go [hunk 1]
```diff
@@ -88,6 +93,37 @@ func (ts *TaskSpec) Validate() *apis.FieldError {
...
+	for _, s := range steps {
+		if s.Image == "" {
+			return apis.ErrMissingField("Image")
+		}
+
+		if s.Name == "" {
+			continue
+		}
+		if _, ok := names[s.Name]; ok {
+			return apis.ErrMultipleOneOf("name")
```
Comment (ImJasonH) [inline pkg/apis/pipeline/v1alpha1/task_validation.go]: I think `ErrMultipleOneOf` is meant to signal that multiple mutually-exclusive fields are specified at the same time. Like a `TaskRun` specifying a `TaskSpec` and a `TaskRef`.

I'm not sure whether there's a field error we should use here, we can't be the only ones who want to prohibit multiple identical keys in a map-like structure. [`ErrInvalidValue`](https://godoc.org/github.com/knative/pkg/apis#ErrInvalidValue) might be the best bet?

Comment (vdemeester) [reply pkg/apis/pipeline/v1alpha1/task_validation.go]: oh right, that's what was used previously, but I do agree :+1:

File: pkg/reconciler/v1alpha1/taskrun/taskrun.go [hunk 2]
```diff
@@ -330,31 +325,105 @@ func (c *Reconciler) reconcile(ctx context.Context, tr *v1alpha1.TaskRun) error
...
 }
 
-func updateStatusFromBuildStatus(taskRun *v1alpha1.TaskRun, buildStatus buildv1alpha1.BuildStatus) {
-	if buildStatus.GetCondition(duckv1alpha1.ConditionSucceeded) != nil {
-		taskRun.Status.SetCondition(buildStatus.GetCondition(duckv1alpha1.ConditionSucceeded))
-	} else {
-		// If the buildStatus doesn't exist yet, it's because we just started running
+func updateStatusFromPod(taskRun *v1alpha1.TaskRun, pod *corev1.Pod) {
+	fmt.Println("taskrun.Status", taskRun.Status)
+	fmt.Println("taskrun.Status.ConditionSucceed", taskRun.Status.GetCondition(duckv1alpha1.ConditionSucceeded))
```
Comment (bobcatfish) [inline pkg/reconciler/v1alpha1/taskrun/taskrun.go]: these lines might have been included by accident? :innocent:

File: pkg/reconciler/v1alpha1/taskrun/taskrun.go [hunk 3]
```diff
@@ -330,31 +325,105 @@ func (c *Reconciler) reconcile(ctx context.Context, tr *v1alpha1.TaskRun) error
...
+	taskRun.Status.StartTime = &pod.CreationTimestamp
+	taskRun.Status.PodName = pod.Name
 
 	taskRun.Status.Steps = []v1alpha1.StepState{}
-	for i := 0; i < len(buildStatus.StepStates); i++ {
-		state := buildStatus.StepStates[i]
-		taskRun.Status.Steps = append(taskRun.Status.Steps, v1alpha1.StepState{
-			ContainerState: *state.DeepCopy(),
+	for _, s := range pod.Status.ContainerStatuses {
+		if s.State.Terminated != nil {
```
Comment (bobcatfish) [inline pkg/reconciler/v1alpha1/taskrun/taskrun.go]: why do we only update the step status if the pod has been terminated? (i think i might be reading this wrong!)

File: pkg/reconciler/v1alpha1/taskrun/taskrun.go [hunk 4]
```diff
@@ -330,31 +325,105 @@ func (c *Reconciler) reconcile(ctx context.Context, tr *v1alpha1.TaskRun) error
...
+			})
+		}
+	}
+
+	switch pod.Status.Phase {
+	case corev1.PodRunning:
+		taskRun.Status.SetCondition(&duckv1alpha1.Condition{
+			Type:   duckv1alpha1.ConditionSucceeded,
+			Status: corev1.ConditionUnknown,
+			Reason: "Building",
```
Comment (bobcatfish) [inline pkg/reconciler/v1alpha1/taskrun/taskrun.go]: maybe `Running` instead of `Building`?

File: pkg/reconciler/v1alpha1/taskrun/taskrun.go (hunk 2 above)
Comment (vdemeester) [reply pkg/reconciler/v1alpha1/taskrun/taskrun.go]: ahh damn I forgot to remove it :sweat_smile: 
#jedi-power-you-never-saw-those-lines

File: pkg/reconciler/v1alpha1/taskrun/taskrun.go (hunk 3 above)
Comment (vdemeester) [reply pkg/reconciler/v1alpha1/taskrun/taskrun.go]: oh you are right, I think I messed up, I'll update

File: pkg/reconciler/v1alpha1/taskrun/taskrun.go (hunk 4 above)
Comment (vdemeester) [reply pkg/reconciler/v1alpha1/taskrun/taskrun.go]: I didn't change what we had before when using `BuildStatus` but yeah I agree.

**Comment by vdemeester** [comment]
> One thought, do you think (if we had coverage mesasurements haha) that the coverage would go up or down with this PR? (or be the same cuz we were using build logic before anyway that maybe wasnt fully covered?) It seems like _maybe_ there is some potential for more unit test coverage

That's a very good point :angel: I'll do some manual measure to make sure i don't lower the coverage.

**Comment by vdemeester** [comment]
For coverage (only changed packages):

| Package                              | Before | After | status |
| ------------------------------------ | ------ | ----- | ------ |
| `pkgs/apis/pipeline/v1alpha1`        | 32.5   | 32.8  | 😄      |
| `pkgs/reconciler/v1alpha1/taskrun`   | 81.1   | 74.2  | 😓      |
| `pkgs/reconciler/v1alpha1/resources` | 89.0   | 87.9  | 😅      |

**Comment by pivotal-nader-ziada** [comment]
@vdemeester are you still going to add tests?

**Comment by vdemeester** [comment]
@pivotal-nader-ziada I'm thinking about it :sweat_smile: I'm just running out of time today :sweat_smile:

**Comment by vdemeester** [comment]
Added some more test, coverage update 

| Package                              | Before | After | status |
| ------------------------------------ | ------ | ----- | ------ |
| `pkgs/apis/pipeline/v1alpha1`        | 32.5   | 32.8  | 😄      |
| `pkgs/reconciler/v1alpha1/taskrun`   | 81.1   | 82.7  | 😄      |
| `pkgs/reconciler/v1alpha1/resources` | 89.0   | 87.9  | 😅      |

**Comment by pivotal-nader-ziada** [comment]
/lgtm

Looking forward to the follow up PR :)

**Pull request merged**